cmd/satellite/main.go

```diff
@@ -1,63 +1,105 @@
 package main
 
 import (
+	"encoding/json"
 	"flag"
 	"log/slog"
 	"os"
 	"time"
 
+	"github.com/gpuctl/gpuctl/internal/config"
 	"github.com/gpuctl/gpuctl/internal/femto"
 	"github.com/gpuctl/gpuctl/internal/gpustats"
 	"github.com/gpuctl/gpuctl/internal/uplink"
 )
 
 func main() {
-	fakeGpus := flag.Bool("fakegpu", false, "Use fake GPU data")
-
-	// don't define flags below here
-	flag.Parse()
-	// don't access flags above here
+	flags := parseProgramFlags()
 
 	log := slog.Default()
 
-	log.Info("Starting satellite", "fakegpu", *fakeGpus)
+	log.Info("Starting satellite", "fakegpu", flags.fakeGPUs)
 
 	host, err := os.Hostname()
+
 	if err != nil {
 		log.Error("failed to get hostname", "err", err)
 		return
 	}
+
 	log.Info("got hostname", "hostname", host)
 
+	satellite_configuration, err := config.GetClientConfiguration("config.toml")
+
+	if err != nil {
+		log.Error("Failed to get satellite configuration from toml configuration file", "err", err)
+	}
+
 	s := satellite{
 		// TODO: Make this configurable
-		gsAddr: "http://localhost:8080",
+		gsAddr: config.GenerateAddress(satellite_configuration.Groundstation.Hostname, satellite_configuration.Groundstation.Port),
 		// we assume hostnames don't change during the program's runtime
 		hostname: host,
 	}
 
-	var hndlr gpustats.GPUDataSource
-	if *fakeGpus {
-		hndlr = gpustats.FakeGPU{}
-	} else {
-		hndlr = gpustats.NvidiaGPUHandler{}
-	}
+	hndlr := setGPUHandler(flags.fakeGPUs)
+
+	go func() {
+		for {
+			log.Info("Sending heartbeat")
+			err := s.sendHeartBeat()
+
+			if err != nil {
+				log.Error("failed to send heartbeat", "err", err)
+			}
+
+			// TODO: testing only, should not send packets this frequently?
+			time.Sleep(time.Duration(satellite_configuration.Satellite.HeartbeatInterval))
+		}
+	}()
+
+	go func() {
+		backlog, _ := recoverState(satellite_configuration.Satellite.Cache)
 
-	for {
-		log.Info("Sending heartbeat")
-		err := s.sendHeartBeat()
-		if err != nil {
-			log.Error("failed to send heartbeat", "err", err)
+		for stat := range backlog {
+			err := s.sendGPUStatus(backlog[stat])
+
+			if err != nil {
+				log.Error("failed to send status", "err", err)
```

Comment:
Would it be possible to distinguish this logging error message to specify it's a message from the backlog? At the moment, it's the same error as if the backlog was empty and the first message sent failed.